app/controllers/course/survey/responses_controller.rb

```diff
@@ -24,20 +24,36 @@ def create
   end
 
   def show
-    render 'course/survey/surveys/index'
+    respond_to do |format|
+      format.html { render 'course/survey/surveys/index' }
+      format.json { render_response_json }
+    end
   end
 
   def edit
-    @response.build_missing_answers_and_options
-    if @response.save
-      render_response_json
-    else
-      head :internal_server_error
+    raise CanCan::AccessDenied if cannot?(:submit, @response) && cannot?(:modify, @response)
```

Comment:
Just to check, is there any cases that student can modify but cannot submit survey ?  Because I am thinking maybe we only need one `modify` ability.